src/cluster_legacy.c

```diff
@@ -1191,6 +1191,20 @@ void clusterInitLast(void) {
     }
 }
 
+/* Called when a cluster node receives SHUTDOWN. */
+void clusterHandleServerShutdown(void) {
+    /* The error logs have been logged in the save function if the save fails. */
+    serverLog(LL_NOTICE, "Saving the cluster configuration file before exiting.");
+    clusterSaveConfig(1);
+
+#if !defined(__sun)
+    /* Unlock the cluster config file before shutdown, see clusterLockConfig. */
+    if (server.cluster_config_file_lock_fd != -1) {
+        flock(server.cluster_config_file_lock_fd, LOCK_UN | LOCK_NB);
+    }
+#endif /* __sun */
+}
```

Comment:
I see this code is moved/refactored from `finishShutdown` but this is technically not needed because all fds will be automatically closed on process termination. Also I think we should continue holding the lock until the very end as I feel that is more consistent with the original intent. 
```suggestion
```

Reply:
yes, it is not technically needed, it is a good habit to do this, clean up proactively as much as possible and make them appear in pairs. Right, i now move it to very end.

Reply:
> it is a good habit to do this,

I would say in this case, symmetry is not a good thing and adds complexity (since we have to worry about future changes breaking this again). The easiest (and correct) thing to do is to just leak the fd and we will never have to worry about this ordering issue.

Reply:
it was added in https://github.com/redis/redis/pull/10912, i kind of forgot the background at that time, i am posting it here first

Reply:
We should add a code comment about it so nobody else will try to delete it in the future.

Reply:
Thanks for the context! An explicit release of this lock makes more sense to me now (and the idea is to not wait for the OS to run down all open fds). BTW, we could just simplify close the fd if it is valid and then we don't need the compile time check plus the matching unlock call, i.e., less code. We should definitely note down this use case as part of this PR too. I am all good here. 

On a side note, I am still curious to know the use case that prompts the lock of the nodes.conf file in the first place. I will try digging into it some other day but in case you guys happen to have the context too...

Reply:
the lock context is in here: https://github.com/redis/redis/commit/db06108bc17205f29e363b5ba1664e1b9dbd9afc

currently we will lock the nodes.conf but we will not lock the valkey.conf, there is a bit of inconsistency here, I guess we will need to lock the valkey.conf later as well

Reply:
> BTW, we could just simplify close the fd if it is valid and then we don't need the compile time check plus the matching unlock call, i.e., less code.

i think we can keep it, i am worried there might be some compilation errors since we will call flock without the `__sun` hint

Reply:
I was suggesting just calling `close`, which will release the lock.

```suggestion
    if (server.cluster_config_file_lock_fd != -1) {
        close(server.cluster_config_file_lock_fd);
    }
```

Reply:
ohh, i got your point. but i still recommend using flock, for literal correctness and also to avoid other potential cross-platform issues

Reply:
Do you have the context? That is not my understanding but I could definitely miss things, as just proved above :-)

To be clear, this is not a blocker to me.

Reply:
i don't have much context on this, i am actually not familiar with cross-platform thing, there is one i think may be related: https://github.com/redis/redis/pull/11327

Reply:
Interesting that Solaris-derived OSes don't support flock(). In that issue, it's mentioned that fcntl can be used for locking on Solaris family of OSes. Maybe it should be done, if we want to support Solaris. I don't know if anyone is using Solaris but we have some code for it. I found an article about how to use fcntl instead of flock: https://www.perkin.org.uk/posts/solaris-portability-flock.html

Reply:
Yeah, `fcntl` sounds like the way to go for Solaris if we would like to retain parity on Solaris. Regardless though, I still think `close` will take care of the unlocking.